BaseTools/Scripts/GitHub.py

```diff
@@ -0,0 +1,160 @@
+        maintainer_file_path (str): The maintainer file path.
+
+    Returns:
+        List[str]: A list of GitHub usernames.
+    """
+
+    out_stream_buffer = StringIO()
+    cmd_ret = RunCmd(
+        "python",
+        f"{maintainer_file_path} -g",
+        workingdir=workspace_path,
+        outstream=out_stream_buffer,
+        logging_level=logging.INFO,
+    )
+    if cmd_ret != 0:
+        print(
+            f"::error title=Reviewer Lookup Error!::Error calling GetMaintainer.py: [{cmd_ret}]: {out_stream_buffer.getvalue()}"
+        )
+        return []
+
+    raw_reviewers = out_stream_buffer.getvalue().strip()
+
+    pattern = r"\[(.*?)\]"
+    matches = re.findall(pattern, raw_reviewers)
+    if not matches:
+        return []
+    matches = list(OrderedDict.fromkeys([match.strip() for match in matches]))
+    return matches
+
+
+def download_gh_file(github_url: str, local_path: str, token=None):
+    """Downloads a file from GitHub.
+
+    Args:
+        github_url (str): The GitHub raw file URL.
+        local_path (str): A local path to write the file contents to.
+        token (_type_, optional): A GitHub authentication token.
+            Only needed for a private repo. Defaults to None.
+    """
+    headers = {}
+    if token:
+        headers["Authorization"] = f"Bearer {token}"
+
+    try:
+        response = requests.get(github_url, headers=headers)
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        print(
+            f"::error title=HTTP Error!::Error downloading {github_url}: {response.reason}"
+        )
+        return
+
+    with open(local_path, "w", encoding="utf-8") as file:
+        file.write(response.text)
+
+
+def add_reviewers_to_pr(
+    token: str, owner: str, repo: str, pr_number: str, user_names: List[str]
+):
+    """Adds the set of GitHub usernames as reviewers to the PR.
+
+    Args:
+        token (str): The GitHub token to use for authentication.
+        owner (str): The GitHub owner (organization) name.
+        repo (str): The GitHub repository name (e.g. 'edk2').
+        pr_number (str): The pull request number.
+        user_names (List[str]): List of GitHub usernames to add as reviewers.
+    """
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Accept": "application/vnd.github.v3+json",
+    }
+    pr_author_url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pr_number}"
+    url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pr_number}/requested_reviewers"
+
+    response = requests.get(pr_author_url, headers=headers)
+    if response.status_code != 200:
+        print(f"::error title=HTTP Error!::Error getting PR author: {response.reason}")
+        return
+    pr_author = response.json().get("user").get("login").strip()
+    while pr_author in user_names:
+        user_names.remove(pr_author)
+    data = {"reviewers": user_names}
+    response = requests.post(url, json=data, headers=headers)

```

Comment:
This algorithm is incomplete.  Changes to files in a PR may require adding and removing reviewers.  ALso, additional reviewers may be manually added by community members and those manually added need to be preserved across a PR change.

Reply:
A more complete algorithm example is shown in link below.  It handles users, teams, and manually added reviews.

 https://github.com/mdkinney/github-action-assign-reviewers/blob/9fa7a68d3edc08f973c7d9d5d43356a67f5b1c1a/src/AssignReviewers.py#L286

Reply:
- This will add new reviewers not already in the PR on updates.
- It will also skip adding reviewers already in the PR (not allowed).

Personally, I didn't see the need to remove reviewers so that was intentional. That was meant to be a simple "automatic reviewer adder" not "automatic reviewer manager". Once someone is added to a PR, even if the code moves it's likely they'd should still be aware, and I didn't want to unexpectedly remove their visibility into the evolution of the change. I leaned toward that being a decision best made by a human and the automation to be strictly additive.

Teams cannot be added with the `GITHUB_TOKEN`. I didn't see that as an issue because reviewers are specified individually in this approach (using GitMaintainer.py output) and a GitHub team would not attempted to be added.

Please let me know if there's an important case I missed.

Reply:
I tested a PR that had a reviewer manually added (not part of the automatic list) and they remained on the PR after synchronization.

Reply:
I agree with the current implementation.  Given that this script and action is intended to address the immediate shortcomings of the process and not the full long-term solution, management of reviewers isn't needed.  The PUSH label is still required and the maintainer still has an active role in the entire process. 

Someday if we can get to code owners and pr process that is merged once policy is met there will need to be a more complete process/automation.

Reply:
I agree with this solution now and can improve later.